File: 21_valid_parenthesis.java
class Solution {
    public boolean isValid(String s) {
      HashMap<Character,Character> map = new HashMap<Character,Character>(){{
        put('}','{');
        put(']','[');
        put(')','(');
      }};
      Stack<Character>stack = new Stack<>();

      for(char c : s.toCharArray()) {
        if(!map.containsKey(c))
          stack.push(c);
        else{

          if(stack.isEmpty()) 
            return false;
          else if(map.get(c) == stack.peek())
            stack.pop();
          else
            return false;
        }

      }
      if(stack.isEmpty())
        return true;
      else 
        return false;
    }
}



/*
   JAVA solution 
TIME:  O(N)
SPACE: O(N)
HashMap<Character,Character> map = new HashMap<Character,Character>(){{
put('}','{');
put(']','[');
put(')','(');
}};
Stack<Character>stack = new Stack<>();

for(char c : s.toCharArray()) {
if(!map.containsKey(c))
stack.push(c);
else{

if(stack.isEmpty()) 
return false;
else if(map.get(c) == stack.peek())
stack.pop();
else
return false;
}

}
if(stack.isEmpty())
return true;
else 
return false;
*/

File: python_solutions/21_valid_parenthesis.py
"""
Python 2 Solution 
(Slow)
"""
class Solution(object):
    def isValid(self, s) -> bool:
        """
        :type s: str
        :rtype: bool
        """
        
        dic = {'}':'{',')':'(',']':'['}
        stack = []
        
        for c in s:
            if c not in dic:
                stack.append(c)
            else:
                if not stack:
                    return False
                if dic.get(c) == stack[-1]:
                    stack.pop()
                else:
                    return False
                
        if not stack:
            return True
        else:
            return False
